casper/src/test/scala/coop/rchain/casper/MergeabilityRules.scala

```diff
@@ -446,19 +496,67 @@ trait MergeabilityRules {
+    "!4 P!"     -> HadItsMatch(S0)(P1)(F0, S1),
+    "!4 (P!)"   -> VolatileEvent(S0)(P1, S1)(F0),
+    "!X PX"     -> IncomingCouldMatch(S0)(P_)(Nil),
+    "!4 PX"     -> HadItsMatch(S0)(P_)(F0),
+    "!C PX"     -> HadItsMatch(S0)(P_)(C_),
+    "!X (P!)"   -> PeekedNoMatch(S0)(P1, S1)(Nil),
+    "!X P!"     -> HadItsMatch(S0)(P_)(S1),
+    "(!4) P!"   -> VolatileEvent(S0, F0)(P_)(S1),
+    "(!4) PX"   -> VolatileEvent(S0, F0)(P_)(Nil),
+    "(!C) P!"   -> PersistentCouldMatch(S0, C_)(P_)(S0),
+    "(!C) PX"   -> PersistentNoMatch(S0, C_)(P_)(Nil),
+    "(4!) (P!)" -> VolatileEvent(F_, S0)(P_, S0)(Nil),
+    "(4!) P!"   -> VolatileEvent(F0, S0)(P_)(S1),
+    "(4!) PX"   -> VolatileEvent(F0, S0)(P_)(Nil),
+    "(P!) (P!)" -> VolatileEvent(P0, S0)(P0, S0)(Nil),
+    "(P!) P!"   -> PeekedNoMatch(P0, S0)(P_)(S1),
+    "4! (P!)"   -> PeekedNoMatch(F0)(P1, S1)(S0),
+    // The case below can merge iff peek block is executed as first - conflicts for now
+    "4! P!"     -> CouldMatchSameConflicts(F0)(P_)(S0),
+    "4! P!"     -> CouldMatchSameMerges(F0)(P1)(S0, S1),
+    "4! PX"     -> SamePolarityMerge(F0)(P1)(S0),
+    "4X PX"     -> SamePolarityMerge(F_)(P_)(Nil),
+    "P! (C!)"   -> PersistentNoMatch(P_)(C0, S0)(S1),
+    "P! (P!)"   -> HadItsMatch(S0)(P_)(F0),
+    "P! C!"     -> CouldMatchSameConflicts(P_)(C_)(S0),
+    "P! P!"     -> CouldMatchSameMerges(P0)(P1)(S0, S1),
+    "PX (C!)"   -> PersistentNoMatch(P0)(C_, S0)(Nil),
+    "PX (P!)"   -> PeekedCouldMatch(P0)(P_, S0)(Nil),
+    "PX (P!)"   -> PeekedCouldMatch(P_, S0)(P0)(Nil),
+    "PX C!"     -> SamePolarityMerge(P0)(C_)(S1),
+    "PX CX"     -> SamePolarityMerge(P_)(C_)(Nil),
+    "PX PX"     -> SamePolarityMerge(P_)(P_)(Nil),
+    "!C P!"     -> HadItsMatch(S0)(P_)(C0, S1),
+    "(!4) (P!)" -> VolatileEvent(S0, F_)(P_, S0)(Nil),
+    "!C (P!)"   -> PeekedNoMatch(S0)(P1, S1)(C0),
+    "(P!) (C!)" -> PeekedCouldMatch(P0, S0)(C_, S0)(Nil),
+    "(!C) (P!)" -> PeekedCouldMatch(S0, C_)(P_, S0)(Nil),
+    "(P!) C!"   -> PeekedCouldMatch(P0, S0)(C0)(S1),
+    "(P!) CX"   -> PeekedCouldMatch(P_, S0)(C_)(Nil),
+    "4X (P!)"   -> PeekedCouldMatch(F0)(P_, S0)(Nil),
+    "P! CX"     -> SamePolarityMerge(P0)(C1)(S0),
+    "4X P!"     -> SamePolarityMerge(F0)(P_)(S1),
+    "PX P!"     -> SamePolarityMerge(P0)(P_)(S1),
+    // TODO:
+    "P! P!" -> CouldMatchSameConflicts(P_)(P_)(S0) // conflicts until replay errors are fixed
+  )
+
   private[this] def conflicts(b1: Rho, b2: Rho, base: Rho)(
       implicit file: sourcecode.File,
       line: sourcecode.Line
   ) =
-    randomDiamondConflictCheck(base, b1, b2, numberOfParentsForDiamondTip = 1)
+    checkBothWays(base, b1, b2, numberOfParentsForDiamondTip = 1)
 
   def merges(b1: Rho, b2: Rho, base: Rho)(
       implicit file: sourcecode.File,
       line: sourcecode.Line
   ) =
-    randomDiamondConflictCheck(base, b1, b2, numberOfParentsForDiamondTip = 2)
+    checkBothWays(base, b1, b2, numberOfParentsForDiamondTip = 2)
```

Comment:
can we revert to `randomDiamondConflictCheck`? It's a matter of 2x test time decrease. I'm fine if 'check both ways' stays as a util for future debugging

Reply:
No, I'd rather have tests take a few seconds longer than need to revert something that made it into `dev` only because tests didn't run i a particular order

Reply:
At Google, there were three levels of tests.  All the short tests taken together for a project would complete in a few minutes.  These would get run regularly during development.  Medium tests would take much longer, up to half an hour.  Medium tests would have to be run once before requesting a code review and again before submitting code.  Long tests could take many hours and involve many disparate systems.  They would be run periodically and would alert anyone who submtted code during that interval when they failed that their change might have caused the failure.  Later they were able to do a binary search (like git bisect) on the relevant PRs to find better information about whose code broke the tests.

Consider agreeing as a team on what constitutes a short test and how long running all the short tests should take.

Reply:
What we're discussing is in the range of short tests. Still, a test longer than 10s affects productivity, because each iteration on the tested code results in a distraction.

In the future, we could replace the 'backend' of this test to just involve the 'merge rules' instead of the whole (test) casper code. We'd then generate an event log by running rholang interpereter on the input rho, feed log to the merge rules (hopefully extracted from EstimatorHelper by that time), and check the 'conflicts' result directly from there, rather than by checking how many parents the 'merge' block has.

Reply:
Sounds good